src/python_testing/TC_TSTAT_4_2.py

```diff
@@ -461,7 +465,86 @@ async def test_TC_TSTAT_4_2(self):
+                    scenarioNotPresent = scenario
+                    break
+
+            if scenarioNotPresent is None:
+                logger.info(
+                    "Couldn't run test step 17 since all preset types in PresetScenarioEnum are supported by this Thermostat")
+            else:
+                test_presets = new_presets_with_handle.copy()
+                test_presets.append(cluster.Structs.PresetStruct(presetHandle=NullValue, presetScenario=scenarioNotPresent,
+                                                                 name="Preset", coolingSetpoint=2500, heatingSetpoint=1700, builtIn=False))
+
+                # Send the AtomicRequest begin command
+                await self.send_atomic_request_begin_command()
+
+                await self.write_presets(endpoint=endpoint, presets=test_presets, expected_status=Status.ConstraintError)
+
+                # Clear state for next test.
+                await self.send_atomic_request_rollback_command()
+
+        self.step("18")
+        if self.pics_guard(self.check_pics("TSTAT.S.F08") and self.check_pics("TSTAT.S.A0050") and self.check_pics("TSTAT.S.Cfe.Rsp")):
+
+            # Read the numberOfPresets supported.
+            numberOfPresetsSupported = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.NumberOfPresets)
+
+            # Read the PresetTypes to get the preset scenarios to build the Presets list.
+            presetTypes = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.PresetTypes)
+
+            # Read the Presets to copy the existing presets into our testPresets list below.
+            presets = await self.read_single_attribute_check_success(endpoint=endpoint, cluster=cluster, attribute=cluster.Attributes.Presets)
+
+            # Calculate the length of the Presets list that could be created using the preset scenarios in PresetTypes and numberOfPresets supported for each scenario.
+            totalExpectedPresetsLength = 0
+            for presetType in presetTypes:
+                totalExpectedPresetsLength += presetType.numberOfPresets
+
+            if totalExpectedPresetsLength > numberOfPresetsSupported:
+                testPresets = []
+                for presetType in presetTypes:
+                    scenario = presetType.presetScenario
+
+                    # For each supported scenario, copy all the existing presets that match it, then add more presets
+                    # until we hit the cap on the number of presets for that scenario.
+                    presetsAddedForScenario = 0
+                    for preset in presets:
+                        if scenario == preset.presetScenario:
+                            testPresets.append(preset)
+                            presetsAddedForScenario = presetsAddedForScenario + 1
+
+                    if presetsAddedForScenario < presetType.numberOfPresets:
```

Comment:
```suggestion
                    while presetsAddedForScenario < presetType.numberOfPresets:
```
Otherwise this does not in fact work right.

Reply:
yeah since number of presets is 1 it passed. yikes fixed it